batch/batch/driver/instance_collection/base.py

```diff
@@ -294,19 +294,19 @@ async def check_on_instance(self, instance: Instance):
 
         log.info(f'{instance} vm_state {vm_state}')
 
-        if (
-            instance.state == 'pending'
-            and isinstance(vm_state, (VMStateCreating, VMStateRunning))
-            and vm_state.time_since_last_state_change() > 5 * 60 * 1000
-        ):
-            log.exception(f'{instance} (state: {str(vm_state)}) has made no progress in last 5m, deleting')
-            await self.call_delete_instance(instance, 'activation_timeout')
-        elif isinstance(vm_state, VMStateTerminated):
+        # Cases are mutually exclusive and therefore order-independent
+        if instance.state == 'pending' and isinstance(vm_state, (VMStateCreating, VMStateRunning)):
+            if vm_state.time_since_last_state_change() > 5 * 60 * 1000:
+                log.exception(f'{instance} (state: {str(vm_state)}) has made no progress in last 5m, deleting')
+                await self.call_delete_instance(instance, 'activation_timeout')
+        elif instance.state in ('pending', 'active') and isinstance(vm_state, VMStateTerminated):
             log.info(f'{instance} live but stopping or terminated, deactivating')
             await instance.deactivate('terminated')
         elif instance.state == 'inactive':
             log.info(f'{instance} (vm_state: {vm_state}) is inactive, deleting')
             await self.call_delete_instance(instance, 'inactive')
+        elif instance.state == 'deleted' and not isinstance(vm_state, VMStateTerminated):
```

Comment:
This is great! I'm glad you formalized the cases. Is it worth adding asserts at the end or no? Something like

```python3
if instance.state == 'active':
     assert is instance(vm_state, VMStateRunning)
```

Actually, I'm a little worried about doing that because I think the OMS update state is treated as VMStateCreating right now.

Reply:
I believe this permits active with Creating or Running, but ya idk how the OMS updates play into this situation.

Reply:
Happy to add asserts for the other cases, there's also the `UnknownVMState` which just gets silently propagated, didn't quite know how to handle that or how much attention to give it. It seems kind of odd to me for `get_vm_state` to return a value in that situation but I'm just not very clued in to this part of the codebase.

Reply:
Unknown VM state is literally that we don't know how to interpret the instance state from the cloud provider and it doesn't match one of the known states we know how to interpret

Reply:
I think asserts would be good if it's not too onerous. I feel like we've messed up this part of the code base multiple times now and it's always been hard to easily verify it has the correct behavior.

Reply:
Will add asserts then